File: packages/sdk/plugin/guard.ts
/**
 * Plugin Guard - Provides stability guarantees for plugin execution.
 *
 * NOT a security sandbox — plugins are considered trusted.
 * Responsibilities:
 * - Timeout enforcement (prevents infinite loops from blocking the engine)
 * - Error isolation (plugin errors don't propagate to the engine)
 */

export interface PluginGuardOptions {
  /** Max execution time per plugin lifecycle hook (ms). 0 = no limit, default 10000. */
  timeout: number;
}

/**
 * Plugin Guard - Executes plugin functions with stability guarantees.
 *
 * This is NOT a security sandbox. It provides:
 * 1. Timeout enforcement — prevents plugin bugs from blocking the engine
 * 2. Error isolation — plugin errors are wrapped, not propagated raw
 */
export class PluginGuard {
  constructor(private options: PluginGuardOptions) {}

  /**
   * Execute a plugin function with timeout and error isolation.
   * If the function times out, a PluginGuardError is thrown.
   * If the function throws, the error is wrapped in a PluginGuardError
   * to prevent plugin errors from propagating directly.
   */
  async execute<T>(pluginId: string, fn: () => Promise<T>): Promise<T> {
    try {
      if (this.options.timeout > 0) {
        return await this.executeWithTimeout(pluginId, fn);
      }
      return await fn();
    } catch (error) {
      // Wrap the error to isolate plugin failures
      throw new PluginGuardError(
        `Plugin '${pluginId}' execution failed: ${error instanceof Error ? error.message : String(error)}`,
        pluginId,
        error instanceof Error ? error : undefined,
      );
    }
  }

  /**
   * Execute a function with a timeout.
   */
  private async executeWithTimeout<T>(pluginId: string, fn: () => Promise<T>): Promise<T> {
    return new Promise<T>((resolve, reject) => {
      const timer = setTimeout(() => {
        reject(new PluginGuardError(
          `Plugin '${pluginId}' timed out after ${this.options.timeout}ms`,
          pluginId,
        ));
      }, this.options.timeout);

      fn()
        .then(result => {
          clearTimeout(timer);
          resolve(result);
        })
        .catch(error => {
          clearTimeout(timer);
          reject(error);
        });
    });
  }
}

/**
 * Error thrown when a plugin guard violation occurs.
 */
export class PluginGuardError extends Error {
  constructor(
    message: string,
    public readonly pluginId: string,
    public override readonly cause?: Error,
  ) {
    super(message);
    this.name = 'PluginGuardError';
  }
}

File: crates/wf-plugin/src/guard.rs
use std::time::Duration;

use futures::FutureExt;
use tokio::time::timeout;

use crate::error::{PluginError, PluginResult};

/// Plugin Guard - Provides stability guarantees for plugin execution.
///
/// This is NOT a security sandbox - plugins are considered trusted.
/// Responsibilities:
/// - Timeout enforcement (prevents infinite loops from blocking the engine)
/// - Panic isolation (a panicking plugin hook does not tear down the engine)
pub struct PluginGuard {
    timeout: Duration,
}

impl PluginGuard {
    pub fn new(timeout_ms: u64) -> Self {
        Self {
            timeout: Duration::from_millis(timeout_ms),
        }
    }

    /// Execute a plugin hook with timeout and panic isolation: a panic inside
    /// the future is caught and reported as [`PluginError::PluginPanic`]
    /// instead of propagating to the engine.
    pub async fn execute<F, T>(&self, plugin_id: &str, f: F) -> PluginResult<T>
    where
        F: std::future::Future<Output = PluginResult<T>> + Send,
    {
        let guarded = async {
            if self.timeout.as_millis() > 0 {
                timeout(self.timeout, f)
                    .await
                    .map_err(|_| PluginError::Timeout {
                        plugin_id: plugin_id.to_owned(),
                    })?
            } else {
                f.await
            }
        };
        match std::panic::AssertUnwindSafe(guarded).catch_unwind().await {
            Ok(result) => result,
            Err(_panic) => {
                tracing::error!(plugin_id, "plugin hook panicked; recovered by PluginGuard");
                Err(PluginError::PluginPanic {
                    plugin_id: plugin_id.to_owned(),
                })
            }
        }
    }
}

impl Default for PluginGuard {
    fn default() -> Self {
        Self::new(10000)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    async fn ok_future() -> PluginResult<u32> {
        Ok(42)
    }

    async fn error_future() -> PluginResult<u32> {
        Err(PluginError::Internal("boom".into()))
    }

    async fn panicking_future() -> PluginResult<u32> {
        panic!("plugin bug");
    }

    async fn slow_future() -> PluginResult<u32> {
        tokio::time::sleep(Duration::from_millis(200)).await;
        Ok(1)
    }

    #[tokio::test]
    async fn successful_execution_passes_through() {
        let guard = PluginGuard::new(1000);
        assert_eq!(guard.execute("p1", ok_future()).await.unwrap(), 42);
    }

    #[tokio::test]
    async fn plugin_errors_are_propagated() {
        let guard = PluginGuard::new(1000);
        assert!(matches!(
            guard.execute("p1", error_future()).await,
            Err(PluginError::Internal(_))
        ));
    }

    #[tokio::test]
    async fn panic_is_isolated_as_plugin_error() {
        let guard = PluginGuard::new(1000);
        assert!(matches!(
            guard.execute("p1", panicking_future()).await,
            Err(PluginError::PluginPanic { .. })
        ));
    }

    #[tokio::test]
    async fn timeout_is_reported_when_configured() {
        let guard = PluginGuard::new(50);
        assert!(matches!(
            guard.execute("p1", slow_future()).await,
            Err(PluginError::Timeout { .. })
        ));
    }

    #[tokio::test]
    async fn zero_timeout_disables_the_guard() {
        let guard = PluginGuard::new(0);
        assert_eq!(guard.execute("p1", ok_future()).await.unwrap(), 42);
    }
}
